lib/modules/datasource/nuget/v3.ts

```diff
@@ -242,3 +267,31 @@ export async function getReleases(
 
   return dep;
 }
+
+async function getSourceUrlFromNupkg(
+  http: Http,
+  packageName: string,
+  packageVersion: string | null,
+  nupkgUrl: string,
+): Promise<string | undefined> {
+  const cacheDir = await ensureCacheDir(`nuget`);
+  const readStream = http.stream(nupkgUrl);
+  try {
+    const nupkgFile = upath.join(
+      cacheDir,
+      `${packageName}.${packageVersion}.nupkg`,
+    );
+    const writeStream = fs.createCacheWriteStream(nupkgFile);
+    await fs.pipeline(readStream, writeStream);
+    const contentsDir = upath.join(
+      cacheDir,
+      `${packageName}.${packageVersion}`,
+    );
+    await extract(nupkgFile, { dir: contentsDir });
+    const nuspecFile = upath.join(contentsDir, `${packageName}.nuspec`);
+    const nuspec = new XmlDocument(await fs.readCacheFile(nuspecFile, 'utf8'));
+    return nuspec.valueWithPath('metadata.repository@url');
+  } finally {
+    await fs.rmCache(cacheDir);
+  }
+}
```

Comment:
I see a couple of options for caching:

1. cache at the level of HTTP call the mapping for nupkgUrl to nupkgFile
2. cache at the level of this function the mapping from nupkgUrl to sourceUrl
3. cache one level above the mapping from (packageName, packageVersion) to sourceUrl

I'd slightly lean towards option 3, because it also optimizes the existing logic. WDYT?

Reply:
see above